accepted/0000-overrides.md

```diff
@@ -0,0 +1,647 @@
+  "bar@1.x": "2.0.0"
+}
+```
+
+The resulting package tree on disk will have `foo` loading `bar@2.0.0` when
+it does `require('bar')`.  Despite this, it will _not_ be treated as an
+invalid dependency, because `bar@1.x` was valid prior to the override being
+applied.
+
+### String Overrides
+
+When the value of an override rule is a string, it is a replacement
+resolution target for resolutions matching the key.  String values _must_
+be a dependency specifier without a name.  (Aliases are supported using the
+`npm:` alias specifier.)
+
+String overrides within a given overrides object are applied in definition
+order, stopping at the _first_ rule to match.  For example:
+
+```js
+{
+  "overrides": {
+    "y@1": "1.2.3",
+    "y@1.2.3": "2.3.4" // <-- this will never match anything
+  }
+}
+```
+
+In this case, the `y@1.2.3 -> 2.3.4` rule will never be applied, because
+any `y@1` dependency will be written to `1.2.3`, and stop evaluating string
+override rules.
+
+This prevents infinite regresses and loops, and greatly simplifies the
+feature both from an implementation and user experience point of view.
+
+For example:
+
+```json
+{
+  "overrides": {
+    "foo@1.0.0": "2.0.0",
+    "foo@2.0.0": "3.0.0",
+    "foo@3.0.0": "1.0.0"
+  }
+}
+```
+
+In this case, a package that depends on `foo@1.0.0` will instead be given
+`foo@2.0.0`.  A package that depends on `foo@2.0.0` will instead be given
+`foo@3.0.0`.  What it will _not_ do is apply the `foo@1.0.0` override to
+`foo@2.0.0`, and then consider whether any other overrides apply to it, and
+so on.  (In this case, it would create an infinite regress.)
+
+A more realistic and less contrived case where a cascade could be
+_desirable_ would be something like this:
+
+```js
+{
+  "overrides": {
+    "webpack@1.x": "2.x", // <-- maybe 2.7.0, maybe some other 2.x
+    "webpack@2.x": "2.7.0"
+  }
+}
+```
+
+In this case, we are saying that any `webpack@1` dependencies should be
+bumped up to `webpack@2`, and furthermore, that any `webpack@2`
+dependencies should be pinned to `webpack@2.7.0`.
+
+In practice, since rules are applied once and not stacked or cascaded, any
+webpack dependency that would resolve to a version matched by `2.x` will be
+overridden to `2.7.0`.  But, any dependency that resolves to a version
+matched by `1.x` will be set to whichever version happens to be the latest
+`2.x` at the time of installation.
+
+In order to produce the intended behavior described, the user would have to
+either specify it twice:
+
+```json
+{
+  "overrides": {
+    "webpack@1.x": "2.7.0",
+    "webpack@2.x": "2.7.0"
+  }
+}
+```
+
+Or make the dependency matching range wider:
+
+```json
+{
+  "overrides": {
+    "webpack@1.x || 2.x": "2.7.0"
+  }
+}
+```
+
+### Object Overrides
+
+An object value in an overrides object defines a _child rule set_.
+
+If the first match for a given resolution is an object, then the object is
+a new rule set applied to all resolutions down that path in the dependency
+graph.
```

Comment:
> all resolutions down that path in the dependency graph

so in the example below, `bar` may be not a direct dependency of `foo`?

Reply:
Correct.  It would be any instance of `bar` at any point in the dependency graph traversal from any `foo` node.